coderd/database/pubsub/latency.go

```diff
@@ -0,0 +1,82 @@
+package pubsub
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+)
+
+// LatencyMeasurer is used to measure the send & receive latencies of the underlying Pubsub implementation. We use these
+// measurements to export metrics which can indicate when a Pubsub implementation's queue is overloaded and/or full.
+type LatencyMeasurer struct {
+	// Create unique pubsub channel names so that multiple coderd replicas do not clash when performing latency measurements.
+	channel uuid.UUID
+	logger  slog.Logger
+}
+
+// LatencyMessageLength is the length of a UUIDv4 encoded to hex.
+const LatencyMessageLength = 36
+
+func NewLatencyMeasurer(logger slog.Logger) *LatencyMeasurer {
+	return &LatencyMeasurer{
+		channel: uuid.New(),
+		logger:  logger,
+	}
+}
+
+// Measure takes a given Pubsub implementation, publishes a message & immediately receives it, and returns the observed latency.
+func (lm *LatencyMeasurer) Measure(ctx context.Context, p Pubsub) (send float64, recv float64, err error) {
+	var (
+		start        time.Time
+		res          = make(chan float64, 1)
+		subscribeErr = make(chan error, 1)
+	)
+
+	msg := []byte(uuid.New().String())
+	log := lm.logger.With(slog.F("msg", msg))
+
+	go func() {
+		_, err = p.Subscribe(lm.latencyChannelName(), func(ctx context.Context, in []byte) {
+			p := p
+			_ = p
+
+			if !bytes.Equal(in, msg) {
+				log.Warn(ctx, "received unexpected message!", slog.F("in", in))
+				return
+			}
+
+			res <- time.Since(start).Seconds()
+		})
+		if err != nil {
+			subscribeErr <- xerrors.Errorf("failed to subscribe: %w", err)
```

Comment:
I'm guessing you're running the subscribe in a separate goroutine so that if subscribe takes longer than the context timeout you won't block.  However, this leaves us with a race between Publish and Subscribe, which if Publish wins means we won't get the message in subscribe.

I think you need to send on subscribeErr unconditionally and then select on it before calling Publish.